### PR TITLE
Copter: Add a switch option to enable/disable RC_Override

### PR DESCRIPTION
**Pull request by khancyr**
Like a RC trainer mode but with rc_override.

### CODE REVIEW

**Comment by olliw42** [comment]
great !! finally :)

I'm interested in this but do not yet exactly know how this override works and didn't look deeply in the code, but I wonder if do_aux_switch_function() is coming after the override of the channel values or before? I could imagine that it's possible to override ch7&ch8 to enable using the assigned function with override, in which case I would think that do_aux_switch_function() is after the override since it's used also by the other functions. If so this could be counter-productive (dangerous) because one could have overridden the override function and wouldn't be able to regain control from the transmitter anymore. In other words, I somehow would have expected that this channel function would be using a different do_aux_switch_XXXX() than the normal channel functions. Just asking to make sure :).

File: ArduCopter/defines.h [hunk 1]
```diff
@@ -74,6 +74,7 @@ enum aux_sw_func {
     AUXSW_INVERTED  =           43,  // enable inverted flight
     AUXSW_WINCH_ENABLE =        44, // winch enable/disable
     AUXSW_WINCH_CONTROL =       45, // winch control
+    AUXSW_ALLOW_RC_OVERRIDE =   46, // enable RC Override
```
Comment (rmackay9) [inline ArduCopter/defines.h]: sorry to be niggly but can we call it RC_OVERRIDE_ENABLE?

File: ArduCopter/GCS_Mavlink.cpp [hunk 2]
```diff
@@ -761,6 +761,14 @@ void GCS_MAVLINK_Copter::handleMessage(mavlink_message_t* msg)
         // or for complete GCS control of switch position
         // and RC PWM values.
         if(msg->sysid != copter.g.sysid_my_gcs) break;                         // Only accept control from our gcs
+        if (!copter.allow_rc_override) {
+            if (copter.failsafe.rc_override_active) {  // if overrides were active previously, disable them
+                copter.failsafe.rc_override_active = false;
+                int16_t v[8]{0};
+                hal.rcin->set_overrides(v, 8);
```
Comment (rmackay9) [inline ArduCopter/GCS_Mavlink.cpp]: I think we can use clear_overrides() instead of writing the full array.

File: ArduCopter/Copter.h [hunk 3]
```diff
@@ -895,6 +895,7 @@ class Copter : public AP_HAL::HAL::Callbacks {
     void do_aux_switch_function(int8_t ch_function, uint8_t ch_flag);
     void save_trim();
     void auto_trim();
+    bool allow_rc_override = true;
```
Comment (rmackay9) [inline ArduCopter/Copter.h]: I can see why putting allow-rc-overrides variable here makes some sense but it is inconsistent with the rest of Copter.h.  Normally variables and methods are defined in different places.  A better place would be within the ap_t structure or below the control_switch_state or perhaps way down below gndeffect_state,

**Comment by rmackay9** [comment]
A nice solution to allow retaking control from the transmitter!  I've added some small suggestions for changes before we merge.  txs!

**Comment by khancyr** [comment]
OK I made change suggested by Randy.

@olliw42 I agree with your consern. Currently, we don't know which aux switch is used for the function. So, here an rc_override can move the switch, eject itself, and prevent to be use anymore. No way to prevent that currently.

**Comment by khancyr** [comment]
Ideally, this option should only be available on a channel not overridable and active only if an rc transmitter is present.

**Comment by rmackay9** [comment]
Merged!

Sad;u, I've just realised that we also need to update CH9_OPT, CH10_OPT, CH11_OPT, CH12_OPT parameter descriptions to add the new option.

I think rc-overrides can only override channel 1 ~ 8 so when we add this to the wiki, we should make sure to mention that users should only assign this to 9 ~ 12.  We could actually add an arming check as well but I'm not sure it's necessary.

**Comment by olliw42** [comment]
thx for the response
I read it to say that a "shutting out" is in fact possible
since IMHO this is a critical function, which (I presume) all users will want to fullyy trust, I would think that one should make this impossible by any means, not only through comments in the wiki ... I mean, isn't that the whole purpose of that function, to allow regaining control in case something went wrong
again, I don't know the code well enough, but I could imagine that one might be able to ignore that option for ch7&ch8? with a simple if statement, maybe in read_aux_switches(), something like if( g.ch7_option != RC_OVERRIDE ) read_aux_switch(CH_7, aux_con.CH7_flag, g.ch7_option);
otherwise I would pledge for a non-disableable arming-check

**Comment by khancyr** [comment]
I made another PR with a simple fix for this problem. I assume we will make this better when move all this option in a proper library

**Comment by olliw42** [comment]
:)

**Comment by rmackay9** [comment]
There may be some follow loopholes to fix.  For example, what happens if you've got a transmitter on with the switch so that it disables the rc-overrides.. but then the transmitter is turned off.  Shouldn't it revert to allowing rc-overrides?

**Comment by khancyr** [comment]
Yep, it should renable the override. I didn't find a nice way to detect that we get a transmiter. Anyway, I address those problems in https://github.com/ArduPilot/ardupilot/pull/7607 as you already merge this one. Can we continue there ?

**Comment by geofrancis** [comment]
This should also be in plane, but since it's still to get chx options I don't know how that would work.

**Pull request merged**